File: sorting/merge_sort/merge_sort.go
/**
* Merge Sort Golang
*/
package main

import (
	"fmt"
)

/**
* Merges the left and right depending on whichever is smaller
* If there is no more entries in either left or right, append the rest
*/
func merge(left, right []int) []int {
	output := make([]int, 0, len(left)+len(right))
	for len(left) > 0 || len(right) > 0 {
		if len(left) == 0 {
			return append(output, right...)
		}
		if len(right) == 0 {
			return append(output, left...)
		}
		if left[0] <= right[0] {
			output = append(output, left[0])
			left = left[1:]
		} else {
			output = append(output, right[0])
			right = right[1:]
		}
	}
	return output
}

/**
* If Array has only one entry return
* Otherwise Recursion split it then merge
*/
func mergeSort(in []int) []int {
	if len(in) <= 1 {
		return in
	}
	middle := len(in) / 2
	left := mergeSort(in[:middle])
	right := mergeSort(in[middle:])
	return merge(left, right)
}

func main() {
	i := []int{1, 44, 22, 92, 94, 12, 3323, 52}
    out := mergeSort(i)
	for _, v := range out {
		fmt.Printf("Value: %d\n", v)
	}
}

File: sorting/merge_sort/merge_sort.c
/**
 * Merge Sort
 */
#include <stdio.h>

void merge(int in[], int left, int middle, int right) {
  int i, j, k;
  int lInd = middle - left + 1;
  int rInd = right - middle;
  int L[lInd], R[rInd];
  for (i = 0; i < lInd; i++)
    L[i] = in[left + i];
  for (j = 0; j < rInd; j++)
    R[j] = in[middle + 1 + j];
  i = 0;
  j = 0;
  k = left;

  while (i < lInd && j < rInd) {
    if (L[i] <= R[j]) {
      in[k] = L[i];
      i++;
    } else {
      in[k] = R[j];
      j++;
    }
    k++;
  }
  while (i < lInd) {
    in[k] = L[i];
    i++;
    k++;
  }
  while (j < rInd) {
    in[k] = R[j];
    j++;
    k++;
  }
}

void mergeSort(int in[], int l, int r) {
  if (l < r) {
    int middle = l + (r - l) / 2;
    mergeSort(in, l, middle);
    mergeSort(in, middle + 1, r);
    merge(in, l, middle, r);
  }
}

int main(int argc, char *argv[argc + 1]) {
  int in[] = {99, 33, 66, 11, 2, 64, 213};
  int size = sizeof(in) / sizeof(in[0]);
  mergeSort(in, 0, size - 1);

  for (int i = 0; i < size; i++) {
    printf("Value: %d\n", in[i]);
  }
  return 0;
}
